python/lsst/ctrl/mpexec/mock_task.py

```diff
@@ -0,0 +1,217 @@
+    ButlerQuantumContext,
+    DeferredDatasetRef,
+    InputQuantizedConnection,
+    OutputQuantizedConnection,
+    PipelineTask,
+    PipelineTaskConfig,
+    PipelineTaskConnections,
+)
+from lsst.utils import doImport
+from lsst.utils.introspection import get_full_type_name
+
+from .dataid_match import DataIdMatch
+
+_LOG = logging.getLogger(__name__)
+
+
+class MockButlerQuantumContext(ButlerQuantumContext):
+    """Implementation of ButlerQuantumContext to use with a mock task.
+
+    Parameters
+    ----------
+    butler : `~lsst.daf.butler.Butler`
+        Data butler instance.
+    quantum : `~lsst.daf.butler.Quantum`
+        Execution quantum.
+
+    Notes
+    -----
+    This implementation overrides get method to try to retrieve dataset from a
+    mock dataset type if it exists. Get method always returns a dictionary.
+    Put method stores the data with a mock dataset type, but also registers
+    DatasetRef with registry using original dataset type.
+    """
+
+    def __init__(self, butler: Butler, quantum: Quantum):
+        super().__init__(butler, quantum)
+        self.butler = butler
+
+    @classmethod
+    def mockDatasetTypeName(cls, datasetTypeName: str) -> str:
+        """Make mock dataset type name from actual dataset type name."""
+        return "_mock_" + datasetTypeName
+
+    def _get(self, ref: DatasetRef) -> Any:
+        # docstring is inherited from the base class
+        if isinstance(ref, DeferredDatasetRef):
+            ref = ref.datasetRef
+        datasetType = ref.datasetType
+
+        typeName, component = datasetType.nameAndComponent()
+        if component is not None:
+            mockDatasetTypeName = self.mockDatasetTypeName(typeName)
+        else:
+            mockDatasetTypeName = self.mockDatasetTypeName(datasetType.name)
+
+        try:
+            mockDatasetType = self.butler.registry.getDatasetType(mockDatasetTypeName)
+            ref = DatasetRef(mockDatasetType, ref.dataId)
+            data = self.butler.get(ref)
+        except KeyError:
+            data = super()._get(ref)
+
+        if not isinstance(data, dict):
+            data = {
+                "ref": {
+                    "dataId": {key.name: ref.dataId[key] for key in ref.dataId.keys()},
+                    "datasetType": ref.datasetType.name,
+                },
+                "type": get_full_type_name(type(data)),
+            }
+        if component is not None:
+            data.update(component=component)
+        return data
+
+    def _put(self, value: Any, ref: DatasetRef):
+        # docstring is inherited from the base class
+
+        mockDatasetType = self.registry.getDatasetType(self.mockDatasetTypeName(ref.datasetType.name))
+        mockRef = DatasetRef(mockDatasetType, ref.dataId)
+        value.setdefault("ref", {}).update(datasetType=mockDatasetType.name)
+        self.butler.put(value, mockRef)
+
+        # also "store" non-mock refs
+        self.registry._importDatasets([ref])
```

Comment:
True, result of the mock running will be "true" output datasets (without actual object in datastore) plus matching mock datasets with provenance info in the same output collection. This means that you can use it to generate quantum graph for the next step (as we do in ci_hsc_gen3).

> we won't be able to use the mocks to test pipeline flow when one task does not produce one of its expected outputs with execution butler. The mock system at present can't really be used for that anyway; it doesn't have a hook for not writing only certain outputs. 

This is true, interaction with the execution butler is not going to work, and missing outputs are not in the picture yet. If I get a better idea what is needed for that kind of test, I could implement additional configurable options in the mock task. And I thought about allowing specializations of mock tasks which can be implemented with more complex logic in them.

And it could be that this thing is mostly useful for accelerating ci_hsc_gen3 and for more realistic tests we need DM-30758 style pipelines.